### PR TITLE
IGNITE-17765 Sql. Introduce cache for parsed statements

### PR DESCRIPTION
**Pull request by AMashenkov**
https://issues.apache.org/jira/browse/IGNITE-17765

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/CacheKey.java [hunk 1]
```diff
@@ -74,25 +66,24 @@ public boolean equals(Object o) {
...
 
-        if (!schemaName.equals(cacheKey.schemaName)) {
+        if (catalogVersion != cacheKey.catalogVersion){
             return false;
         }
         if (!query.equals(cacheKey.query)) {
             return false;
         }
-        if (!Objects.equals(contextKey, cacheKey.contextKey)) {
+        if (!query.equals(cacheKey.query)) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/CacheKey.java:75]: you've checked `query` already

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 2]
```diff
@@ -374,121 +389,190 @@ private void registerIndexListener(IndexEvent evt, AbstractIndexEventListener ls
...
+                .logger(LOG)
+                .cancel(queryCancel)
+                .parameters(params)
+                .plannerTimeout(PLANNER_TIMEOUT)
+                .build();
+
+        CompletableFuture<QueryPlan> planFuture = queryCache.get(cacheKey);
+
+        if (planFuture == null) {
+            planFuture = CompletableFuture.supplyAsync(() -> {
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:444]: I think it would be better to specify an executor here since `CompletableFuture.supplyAsync(task)` uses ForkJoin's common pool.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 3]
```diff
@@ -247,7 +261,8 @@ public synchronized void start() {
                 busyLock
         );
 
-        sqlSchemaManager.registerListener(prepareSvc);
+        //TODO IGNITE-17765 Do we need cache clear as cache already aware of catalog versioning.
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:264]: well, we should not refer in TODO to the ticket under which this TODO is being added

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/CacheKey.java [hunk 4]
```diff
@@ -74,25 +65,24 @@ public boolean equals(Object o) {
...
 
-        return Arrays.deepEquals(paramTypes, cacheKey.paramTypes);
+        return Arrays.equals(paramTypes, cacheKey.paramTypes);
     }
 
     @Override
     public int hashCode() {
         int result = schemaName.hashCode();
         result = 31 * result + query.hashCode();
-        result = 31 * result + (contextKey != null ? contextKey.hashCode() : 0);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/CacheKey.java:94]: catalogVersion?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 5]
```diff
@@ -383,121 +398,187 @@ private void registerIndexListener(IndexEvent evt, AbstractIndexEventListener ls
...
         if (session == null) {
-            return CompletableFuture.failedFuture(
+            return failedFuture(
                     new SqlException(SESSION_NOT_FOUND_ERR, format("Session not found [{}]", sessionId)));
         }
 
+        QueryCancel queryCancel = new QueryCancel();
+
+        try {
+            registerQueryCancel(session, new QueryCancel());
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:415]: Why do you need new object in the session?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 6]
```diff
@@ -383,121 +398,187 @@ private void registerIndexListener(IndexEvent evt, AbstractIndexEventListener ls
...
-        InternalTransaction outerTx = context.unwrap(InternalTransaction.class);
+        InternalTransaction outerTx = queryContext.unwrap(InternalTransaction.class);
+        boolean implicitTx = outerTx == null;
 
-        QueryCancel queryCancel = new QueryCancel();
+        // TODO: IGNITE-19497 Switch to Catalog manager and uncomment next lines.
+        long timestamp = outerTx == null ? clock.nowLong() : outerTx.startTimestamp().longValue();
+        // int plannerCatalogVersion = catalogManager.activeCatalogVersion(timestamp);
+        // int schemaId = catalogManager.schema(plannerCatalogVersion).id();
+        // SchemaPlus schema = sqlSchemaManager.schema(schemaName, plannerCatalogVersion);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:430]: Let's not jump to conclusions about how unrelated tickets might look like

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 7]
```diff
@@ -383,121 +398,187 @@ private void registerIndexListener(IndexEvent evt, AbstractIndexEventListener ls
...
+                        int txCatalogVersion = catalogManager.activeCatalogVersion(tx.startTimestamp().longValue());
+
+                        if (implicitTx && plannerCatalogVersion != txCatalogVersion) {
+                            LOG.info("Retry query planning: plannerCatalogVersion={}, txCatalogVersion={}",
+                                    plannerCatalogVersion, txCatalogVersion);
+
+                        // TODO IGNITE-19497 Prevent infinite planning, e.g. by setting max number of tries.
+                            return tx.rollbackAsync()
+                                    .thenComposeAsync(ignore -> querySingle0(sessionId, queryContext, sql, params), taskExecutor);
+                        } */
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:506]: the same, let's remove all unrelated code

**Comment by AMashenkov** [comment]
- Cache plans for original query string as for normalized query string. Cache coherence for original and normalized string is not supported.
- Fixed EXPLAIN to use cached plan
- Tests added
